packages/eas-cli/src/utils/download.ts

```diff
@@ -0,0 +1,160 @@
+  Log.log(infoMessage ? infoMessage : `Downloading file from ${url}...`);
+
+  const downloadProgressBar = new cliProgress.SingleBar(
+    { format: '|{bar}|' },
+    cliProgress.Presets.rect
+  );
+
+  let downloadProgressBarStarted = false;
+  const response = await wrapFetchWithProgress()(
+    url,
+    {
+      timeout: 1000 * 60 * 5, // 5 minutes
+    },
+    (_progress: number, total: number, loaded: number): void => {
+      if (!downloadProgressBarStarted) {
+        downloadProgressBar.start(total, loaded);
+        downloadProgressBarStarted = true;
+      } else if (loaded < total) {
+        downloadProgressBar.update(loaded);
+      } else {
+        downloadProgressBar.stop();
+      }
+    }
+  );
+
+  if (!response.ok) {
+    throw new Error(`Failed to download file from ${url}`);
+  }
+
+  await pipeline(response.body, fs.createWriteStream(outputPath));
+}
+
+export async function downloadAndExtractAppAsync(
+  url: string,
+  applicationExtension: string
+): Promise<string> {
+  const outputDir = await ensureTmpDirExistAsync(path.join(getTmpDirectory(), uuidv4()));
+
+  const tmpArchivePathDir = await ensureTmpDirExistAsync(
+    path.join(getTmpDirectory(), `${uuidv4()}`)
+  );
+  const tmpArchivePath = path.join(tmpArchivePathDir, `${uuidv4()}.tar.gz`);
+
+  await downloadFileWithProgressBarAsync(url, tmpArchivePath, 'Downloading app archive...');
+  await tarExtractAsync(tmpArchivePath, outputDir);
+
+  return await getAppPathAsync(outputDir, applicationExtension);
+}
+
+export async function extractAppFromLocalArchiveAsync(
+  appArchivePath: string,
+  applicationExtension: string
+): Promise<string> {
+  const outputDir = await ensureTmpDirExistAsync(path.join(getTmpDirectory(), uuidv4()));
+
+  await tarExtractAsync(appArchivePath, outputDir);
+
+  return await getAppPathAsync(outputDir, applicationExtension);
+}
+
+async function getAppPathAsync(outputDir: string, applicationExtension: string): Promise<string> {
+  const appFileName = await glob(`*.${applicationExtension}`, {
+    cwd: outputDir,
+    onlyFiles: false,
+  });
+
+  if (appFileName.length === 0) {
+    throw Error('Something went wrong while extracting the app from app archive');
+  }
+
+  return path.join(outputDir, appFileName[0]);
+}
+
+async function tarExtractAsync(input: string, output: string): Promise<void> {
+  try {
+    if (process.platform !== 'win32') {
+      await spawnAsync('tar', ['-xf', input, '-C', output], {
+        stdio: 'inherit',
+      });
+      return;
+    }
+  } catch (error: any) {
+    Log.warn(
+      `Failed to extract tar using native tools, falling back on JS tar module. ${error.message}`
+    );
+  }
+  Log.debug(`Extracting ${input} to ${output} using JS tar module`);
+  // tar node module has previously had problems with big files, and seems to
+  // be slower, so only use it as a backup.
+  await extract({ file: input, cwd: output });
+}
+
+async function ensureTmpDirExistAsync(dirPath: string): Promise<string> {
```

Comment:
- name ensureTmpDirExistAsync implies it's only validating tmp dirs, but this function could work with any dir
- functions named ensure... should not fail if their goal is already achieved, this function will throw if the directory exists
- calling this function is more code than just inlining it, I see that you also return the path, but ir seems like you are doing it only to save a line in a calling code, which is a bit hacky.

I would suggest either inlining this function or writing function `getTmpDirAsync()` that generates name ensures the directory exists and returns it.

also you can use `mkdirp` from `fs-extra` so the function does not fail if directory exists